modules/packages/EpochManager.chpl

```diff
@@ -0,0 +1,1147 @@
+        return n;
+      }
+
+      proc enqueue(newObj : objType) {
+        var n = recycle_node();
+        n.val = newObj;
+
+        // Now enqueue
+        while (true) {
+          var tail = _tail.readABA();
+          var next = tail.next.readABA();
+          var next_node = next.getObject();
+          var curr_tail = _tail.readABA();
+          if (tail == curr_tail) {
+            if (next_node == nil) {
+              if (curr_tail.next.compareExchangeABA(next, n)) {
+                _tail.compareExchangeABA(curr_tail, n);
+                break;
+              }
+            }
+            else {
+              _tail.compareExchangeABA(curr_tail, next_node);
+            }
+          }
+        }
+      }
+
+      proc dequeue() : objType {
+        while (true) {
+          var head = _head.readABA();
+          var head_node = head.getObject();
+          var curr_tail = _tail.readABA();
+          var tail_node = curr_tail.getObject();
+          var next = head.next.readABA();
+          var next_node = next.getObject();
+          var curr_head = _head.readABA();
+
+          if (head == curr_head) {
+            if (head_node == tail_node) {
+              if (next_node == nil) then
+                return nil;
+              _tail.compareExchangeABA(curr_tail, next_node);
+            }
+            else {
+              var ret_val = next_node.val;
+              if (_head.compareExchangeABA(curr_head, next_node)) {
+                retire_node(head_node);
+                return ret_val;
+              }
+            }
+          }
+        }
+
+        return nil;
+      }
+
+      // TODO: Reclaim retired nodes after a while
+      proc retire_node(nextObj : unmanaged node(objType)) {
+        nextObj.val = nil;
+        do {
+          var oldTop = _freeListHead.readABA();
+          nextObj.freeListNext = oldTop.getObject();
+        } while (!_freeListHead.compareExchangeABA(oldTop, nextObj));
+      }
+
+      iter these() : objType {
+        var ptr = _head.read().next.read();
+        while (ptr != nil) {
+          yield ptr.val;
+          ptr = ptr.next.read();
+        }
+      }
+
+      proc peek() : objType {
+        var actual_head = _head.read().next.read();
+        if (actual_head != nil) then
+          return actual_head.val;
+        return nil;
+      }
+
+      proc deinit() {
+        var ptr = _head.read();
+        if (delete_val) {
+          while (ptr != nil) {
+            _head = ptr.next;
+            delete ptr.val;
+            delete ptr;
+            ptr = _head.read();
+          }
+        } else {
+          while (ptr != nil) {
+            _head = ptr.next;
+            delete ptr;
+            ptr = _head.read();
+          }
+        }
+
+        ptr = _freeListHead.read();
+        while (ptr != nil) {
+          var head = ptr.freeListNext;
+          delete ptr;
+          ptr = head;
+        }
+      }
+    }
+  }
+
+  pragma "no doc"
+  module LimboListModule {
```

Comment:
Even if this is not user facing, please add some comments indicating generally what it is / how it is used in the context of EBR. These comments are important to help future developers in maintaining the code.

Reply:
Gotcha, makes sense. We'll add more documentation.